synfig-studio/src/gui/main.cpp

```diff
@@ -68,75 +67,43 @@ using namespace studio;
 int main(int argc, char **argv)
 {
 
-#ifdef _WIN32
-	// to be able to open files whose name is not latin (eg. arabic)
-	class ArgVGuard {
-		char **modified_argv;
-	public:
-		ArgVGuard(char ***argv) { modified_argv = *argv = g_win32_get_command_line(); }
-		~ArgVGuard() { g_strfreev(modified_argv); }
-	} argv_guard(&argv);
- #endif
-
-#ifdef _WIN32
-	if (consoleOptionEnabled(argc, argv))
-	{
-		redirectIOToConsole();
-	}
-	else
-	{
-		// QuickHack: to avoid strange bug with stderr
-		freopen("NUL", "w", stdout);
-		freopen("NUL", "w", stderr);
-		freopen("NUL", "r", stdin);
-		ios::sync_with_stdio();
-	}
-#endif
-
-	String binary_path = synfig::get_binary_path(String(argv[0]));
+	const String binary_path = synfig::get_binary_path(String(argv[0]));
+	const String rootpath = etl::dirname(etl::dirname(binary_path));
 	
 #ifdef ENABLE_NLS
 	String locale_dir;
-	locale_dir = etl::dirname(etl::dirname(binary_path))+ETL_DIRECTORY_SEPARATOR+"share"+ETL_DIRECTORY_SEPARATOR+"locale";
+	locale_dir = rootpath+ETL_DIRECTORY_SEPARATOR+"share"+ETL_DIRECTORY_SEPARATOR+"locale";
 	setlocale(LC_ALL, "");
 	bindtextdomain(GETTEXT_PACKAGE,  Glib::locale_from_utf8(locale_dir).c_str() );
 	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
 	textdomain(GETTEXT_PACKAGE);
 #endif
 	
-	{
-		SmartFILE file(IPC::make_connection());
-		if(file)
-		{
-			cout << endl;
-			cout << "   " << _("synfig studio is already running") << endl << endl;
-			cout << "   " << _("the existing process will be used") << endl << endl;
-
-			// Hey, another copy of us is open!
-			// don't bother opening us, just go ahead and
-			// tell the other copy to load it all up
-			if (argc>1)
-				fprintf(file.get(),"F\n");
-
-			while(--argc)
-				if((argv)[argc] && (argv)[argc][0]!='-')
-					fprintf(file.get(),"O %s\n",etl::absolute_path((argv)[argc]).c_str());
-
-			fprintf(file.get(),"F\n");
-
-			return 0;
-		}
-	}
-
 	cout << endl;
 	cout << "   " << _("synfig studio -- starting up application...") << endl << endl;
 
 	SYNFIG_EXCEPTION_GUARD_BEGIN()
-	studio::App app(etl::dirname(binary_path), &argc, &argv);
+	
+	Glib::RefPtr<studio::App> app = studio::App::instance();
+	bool is_remote = true;
+
+	app->signal_startup().connect([app, rootpath, &is_remote]() {
+		app->init(rootpath);
+		is_remote = false;
+	});
+
+	int exit_code = app->run(argc, argv);
+
+	// app->is_remote() give us a nasty warning
+	if (is_remote) {
```

Comment:
> app->is_remote() give us a nasty warning

Which one? I see only standard message after changing this to `if (app->is_remote())`.

Reply:
I suppose that `app` is already closed at that line, so `app` is not registered anymore.
It makes sense: once app shuts down, it should de-register itself.

Reply:
Can you describe how you are testing this? 
In my case I launch Synfig and try to launch it again from another console.

Reply:
Same way.

1. Now I edited the code to be like this (for my previous post I did it in another way):
![Captura de tela de 2022-01-05 13-56-05](https://user-images.githubusercontent.com/1323987/148257309-213b4474-9e4b-4fb9-90a7-2284ce1a002e.png)
2. Launched it normally.
3. While it is open, I launch it again from another terminal and I got this:
![Captura de tela de 2022-01-05 13-57-16](https://user-images.githubusercontent.com/1323987/148257523-8609dc05-007c-4a97-975f-4f98ae8bac7d.png)
4. Close primary instance and I got this:
![Captura de tela de 2022-01-05 13-58-55](https://user-images.githubusercontent.com/1323987/148257687-4b2c3abb-437b-4812-852c-9ba1199a2083.png)

Reply:
Ok. This is what I found:
1. `Gio::Application::is_remote()` [docs](https://developer-old.gnome.org/glibmm/unstable/classGio_1_1Application.html#afb0429cbaaaf8bc48f461a90507fc157)
> **bool Gio::Application::is_remote() const**
> If application is remote then it means that another instance of application already exists (the 'primary' instance). Calls to perform actions on application will result in the actions being performed by the primary instance.

> The value of this property cannot be accessed before g_application_register() has been called. See g_application_get_is_registered().

2. `g_application_register()` [docs](https://docs.gtk.org/gio/method.Application.register.html)
> The GApplication::startup signal is emitted if registration succeeds and application is the primary instance (including the non-unique case).

3. https://gitlab.gnome.org/GNOME/glib/-/blob/HEAD/gio/tests/gapplication-example-actions.c

Conclusions:
1. `Gio.Application` registers itself after startup, and a good point for check is the `startup` signal.
2. We can't use `startup` signal, because it is only emitted for the primary instance, so we have to use a different signal or a different approach. For example, the current one.
3. After checking `gapplication-example-actions.c` I made a change that work without error.

```
	Glib::RefPtr<studio::App> app = studio::App::instance();

	app->signal_startup().connect([app, rootpath]() {
		app->init(rootpath);
	});

	app->register_application();
	if (app->is_remote()) {
		std::cout << endl;
		std::cout << "   " << _("synfig studio is already running") << std::endl << std::endl;
		std::cout << "   " << _("the existing process will be used") << std::endl << std::endl;
	}
```

Reply:
It seems to launch the app once more on quitting (Splash Screen appears and all) before do quit.

Reply:
I didn’t notice it. What should I do to reproduce this?

Reply:
I launched it.
In other terminal, I tried to relaunch it - everything worked fine: it said it is already running, popped up the already open Synfig Studio window and exits.
I close Synfig Studio: splash screen appeared, and said loading all that stuff as if it was starting Studio again, but closes at end.

Reply:
What's strange. Is it reproduced if you not trying to relaunch it?
Can you push these changes to a new branch so I can check them?

Reply:
Hm, only reproduced on Debian 12. I will take a look.

Reply:
I think this is happening because your code looks like this: 
```
	app->signal_startup().connect([app, rootpath]() {
		app->init(rootpath);
	});

	int exit_code = app->run(argc, argv);

	app->register_application();
	if (app->is_remote()) {
		std::cout << endl;
		std::cout << "   " << _("synfig studio is already running") << std::endl << std::endl;
		std::cout << "   " << _("the existing process will be used") << std::endl << std::endl;
	}
```

but it should be:
```
	app->signal_startup().connect([app, rootpath]() {
		app->init(rootpath);
	});

	app->register_application();
	if (app->is_remote()) {
		std::cout << endl;
		std::cout << "   " << _("synfig studio is already running") << std::endl << std::endl;
		std::cout << "   " << _("the existing process will be used") << std::endl << std::endl;
	}

	int exit_code = app->run(argc, argv);
```

Reply:
Ouch.

Well these are the patches Debian team apply themselves to gtk3 in debian 12 (testing/Bookworm): https://sources.debian.org/patches/gtk+3.0/3.24.31-1/

Reply:
gtkmm has no patches: https://sources.debian.org/patches/gtkmm3.0/3.24.5-1/

Reply:
> Hm, only reproduced on Debian 12. I will take a look.

I mean this is not a Debian issue, it is related to the newer GTK. 
Can you check my assumption about `app->run()` location?

Reply:
I answered you before Github updates your proposal XD
It works! I made it wrong, sorry :(

Reply:
Done :)